globus_sdk/config/environments.py

```diff
@@ -0,0 +1,85 @@
+import logging
+import typing
+
+log = logging.getLogger(__name__)
+
+
+class EnvConfig:
+    envname: str
+    domain: str
+    no_dotapi: typing.List[str] = ["auth"]
+
+    # this same dict is inherited (and therefore shared!) by all subclasses
+    _registry: typing.Dict[str, typing.Type["EnvConfig"]] = {}
+
+    # this is an easier hook to use than metaclass definition -- register every subclass
+    # in this dict automatically
+    #
+    # as a result, anyone can define
+    #
+    #       class BetaEnv(EnvConfig):
+    #           domain = "beta.foo.bar.example.com"
+    #           envname = "beta"
+    #
+    # and retrieve it with get_config_by_name("beta")
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._registry[cls.envname] = cls
+
+    @classmethod
+    def get_service_url(cls, service):
+        # you can override any name with a config attribute
+        service_url_attr = f"{service}_url"
+        if hasattr(cls, service_url_attr):
+            return getattr(cls, service_url_attr)
+
+        # the typical pattern for a service hostname is X.api.Y
+        # X=transfer, Y=preview.globus.org => transfer.api.preview.globus.org
+        # check `no_dotapi` for services which don't have `.api` in their names
+        if service in cls.no_dotapi:
+            return f"https://{service}.{cls.domain}/"
+        return f"https://{service}.api.{cls.domain}/"
+
+
+# the main interfaces to this module
+#   get_config_by_name("test") -> TestEnvConfig
+def get_config_by_name(envname: str) -> typing.Optional[typing.Type[EnvConfig]]:
```

Comment:
This strikes me as something that would frequently be written as a classmethod: `EnvConfig.by_name(envname)`, and the "hidden" variable use seems to bear that out

Reply:
Yeah, let's go with that and put the `EnvConfig` class into the `__init__.py`. That makes it easier to grab it and write `class DemoEnv(EnvConfig): ...` or any other interesting use-cases.